src/format_config.ml

```diff
@@ -0,0 +1,151 @@
+open! Stdune
+open Import
+open Dune_lang.Decoder
+
+let syntax =
+  Syntax.create ~name:"fmt"
+    ~desc:"integration with automatic formatters"
+    [ (1, 2) ]
+
+type language =
+  | Dialect of string
+  | Dune
+
+let language_to_dyn =
+  let open Dyn.Encoder in
+  function
+  | Dialect name -> constr "dialect" [string name]
+  | Dune -> constr "dune" []
+
+let language = function
+  | "dune" -> Dune
+  | s -> Dialect s
+
+type enabled_for =
+  | Only of language list
+  | All
+
+let languages_in_ext_1_0 =
+  [Dialect "ocaml"; Dialect "reason"]
+
+let languages_in_ext_1_1 =
+  Dune::languages_in_ext_1_0
+
+let enabled_for_field =
+ field_o "enabled_for" (repeat (map ~f:language string))
+
+let enabled_for_field_ext =
+  let+ list_opt = enabled_for_field
+  and+ ext_version = Syntax.get_exn syntax
+  in
+  match list_opt, ext_version with
+  | Some l, _ -> Only l
+  | None, (1, 0) -> Only languages_in_ext_1_0
+  | None, (1, 1) -> Only languages_in_ext_1_1
+  | None, (1, 2) -> All
+  | None, _ ->
+    Code_error.raise "This fmt version does not exist"
+      [("version", Syntax.Version.to_dyn ext_version)]
+
+let enabled_for_to_dyn =
+  let open Dyn.Encoder in
+  function
+  | Only l -> constr "only" (List.map ~f:language_to_dyn l)
+  | All -> string "all"
+
+type t =
+  { loc : Loc.t
+  ; enabled_for : enabled_for
+  }
+
+let to_dyn {enabled_for; loc = _} =
+  let open Dyn.Encoder in
+  record ["enabled_for", enabled_for_to_dyn enabled_for]
+
+let dparse_args =
+  let+ loc = loc
+  and+ enabled_for = fields enabled_for_field_ext
+  in
+  ({loc; enabled_for}, [])
+
+let includes config =
+  match config.enabled_for with
+  | Only l -> List.mem ~set:l
+  | All -> fun _ -> true
+
+let dune2_record_syntax =
+  let+ loc = loc
+  and+ ef = enabled_for_field
+  in
+  let enabled_for = match ef with
+    | Some l -> Only l
+    | None -> All
+  in
+  Some {loc; enabled_for}
+
+let dune2_dec =
+  if_list
+    ~then_:(fields dune2_record_syntax)
+    ~else_:(keyword "disabled" >>= fun () -> return None)
+
+let dune2_default =
+  Some
+    { loc = Loc.none
+    ; enabled_for = All
+    }
+
+let field_dune2 =
+  field_o "formatting" dune2_dec >>| Option.value ~default:dune2_default
+
+let field =
+  Syntax.get_exn Stanza.syntax >>= fun dune_lang_version ->
+  match Syntax.Version.compare dune_lang_version (2, 0) with
+  | Lt -> return None
+  | Gt | Eq ->
+    field_dune2
+
+let loc t = t.loc
+
+exception Delete
+
+module Encoder = struct
+  open Dune_lang.Encoder
+
+  let encode_language = function
+    | Dune -> string "dune"
+    | Dialect d -> string d
+
+  let explicit_langs =
+    function
+    | Only l -> l
+    | All -> raise Delete
```

Comment:
Not a huge deal here since we are in an error case, but in general it is better to use `Exn.raise_notrace` when using an exception to abort a computation.